ios/browser/api/history/brave_history_api.mm

```diff
@@ -0,0 +1,283 @@
+}
+
+- (void)removeAllWithCompletion:(void(^)())completion {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  // Deletes entire History and from all synced devices
+  history_service_->DeleteLocalAndRemoteHistoryBetween(web_history_service_,
+                                                      base::Time::Min(),
+                                                      base::Time::Max(),
+                                                      base::BindOnce([](std::function<void()> completion){
+                                                        completion();
+                                                      }, completion),
+                                                      &tracker_);
+}
+
+- (void)searchWithQuery:(NSString*)query maxCount:(NSUInteger)maxCount
+                                       completion:(void(^)(NSArray<IOSHistoryNode*>* historyResults))completion {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  // Check Query is empty for Fetching all history
+  // The entered query can be nil or empty String
+  BOOL fetchAllHistory = !query || [query length] > 0;
+  std::u16string queryString =
+        fetchAllHistory ? std::u16string() : base::SysNSStringToUTF16(query);
+
+  // Creating fetch options for querying history
+  history::QueryOptions options;
+  options.duplicate_policy =
+      fetchAllHistory ? history::QueryOptions::REMOVE_DUPLICATES_PER_DAY
+                      : history::QueryOptions::REMOVE_ALL_DUPLICATES;
+  options.max_count = fetchAllHistory ? 0 : static_cast<int>(maxCount);
+  options.matching_algorithm =
+      query_parser::MatchingAlgorithm::ALWAYS_PREFIX_SEARCH;
+
+  _query_completion = completion;
+
+  browsing_history_service_->QueryHistory(queryString, options);
+}
+
+#pragma mark - BraveHistoryDriverDelegate
+
+- (void)historyQueryWasCompletedWithResults:
+    (const std::vector<history::BrowsingHistoryService::HistoryEntry>&)results
+                   queryResultsInfo:(const history::BrowsingHistoryService::
+                                         QueryResultsInfo&)queryResultsInfo
+                        continuationClosure:(base::OnceClosure)continuationClosure {
+  
+  if (_query_completion) {
+    continuationClosure.Reset();
+    
+    NSMutableArray<IOSHistoryNode*>* historyNodes = [[NSMutableArray alloc] init];
+
+    for (const auto& result : results) {
+      IOSHistoryNode *historyNode = [[IOSHistoryNode alloc] initWithURL:net::NSURLWithGURL(result.url)
+                                                                  title:base::SysUTF16ToNSString(result.title)
+                                                              dateAdded:[NSDate dateWithTimeIntervalSince1970:
+                                                                            result.time.ToDoubleT()]];
+      [historyNodes addObject:historyNode];
+    }
+
+    _query_completion([historyNodes copy]);
```

Comment:
No need to copy here because the array is local anyway and isn't a stored property :)

Reply:
True!, It was global before my bad.